content/gotraining.slide

```diff
@@ -212,21 +224,265 @@ Execute with
+	2db0f12 fixed two build-breaking issues: + reverted ClassMetadataReadingVisitor to revision 794 + eliminated ConfigurationPostProcessorTests until further investigation determines why it causes downstream tests to fail (such as the seemingly unrelated ClassPathXmlApplicationContextTests)
+	147709f Tweaks to package-info.java files
+	22b25e0 Consolidated Util and MutableAnnotationUtils classes into existing AsmUtils
+	7f96f57 polishing
+
+versus
+
+	$ git log --oneline -5 --author pwebb --before "Sat Aug 30 2014"
+
+	5ba3db6 Fix failing CompositePropertySourceTests
+	84564a0 Rework @PropertySource early parsing logic
+	e142fd1 Add tests for ImportSelector meta-data
+	887815f Update docbook dependency and generate epub
+	ac8326d Polish mockito usage
+
+
+* Recipe for great commit messages
+
+[[https://chris.beams.io/posts/git-commit/]]
+
+- Separate subject from body with a blank line
+- Limit the subject line to 60 characters
+- *Capitalize* the subject line
+- Do not end the subject line with a period
+- Use the *imperative* *mood* in the subject line
+- Wrap the body at 80 characters
+- Use the body to explain what and why vs. how
+- Use `git` `rebase` `-i` `COMMIT_HASH` to rework your commits if necessary
+
+Don't be afraid!
+Try it out on a new, throwaway branch.
+
+
+* Pull Request (PR) recommendations (1/2)
+
+- Choose a meaningful *PR* *title*, it will become your Squash Merge message
+- Fill in *PR* *description* and reference an Issue for instance with #6
+- Review the *Files* *changed* tab - many reviewers start here
+- Review the *Commits* tab - some reviewers start here
+- Review the *Checks* tab - all checks must pass
+
+
+* Pull Request (PR) recommendations (2/2)
+
+- "Think of the reviewer: your code needs to be reviewable and that should be a prime concern when writing your code, commits and PRs." (@camh-anz)
+- Tag your PR Title with `[WIP]` if adjustments are needed
+- Rework fixup commits with `git` `rebase` `-i`  `COMMIT_HASH` and force push
+- Add reviewers and remove `[WIP]` tag when all of the above are met
+- Delete branch once merged, especially when using a single remote (no forks)
+
+
 * Go language fundamentals
 
 
+* Tutorials, references and tools
+
+- [[https://tour.golang.org/][Tour of Go]]
+- [[https://gobyexample.com/][Go by example]]
+- [[https://play.golang.org/][Go playground]]
+- [[https://golangbot.com/learn-golang-series/][Golangbot]]
+- [[https://stackoverflow.com/questions/tagged/go?sort=votes&pageSize=100][Stackoverflow]]
+
+
+* Values and Control Flow
+
+- Values
+- Variables
+- Constants
+- For
+- If/Else
+- Switch
+
+
+* Lab 1
+
+TODO
+
+
+* Iterable data types and Range
+- Arrays
+- Slices
+- Maps
+- Range
+
+
+* Lab 2
+
+TODO
+
+
+* Functions
+
+- Functions
+- Multiple Return Values
+- Variadic Functions
+- Closures
+
+
+* Lab 3
+
+TODO
+
+
+* Data structures and interfaces
+
+- Pointers
+- Structs
+- Methods
+- Interfaces
+
+
+* Lab 4
+
+TODO
+
+
+* Errors
+
+- Errors
+- Defer
```

Comment:
personally i think of defer as not really being related to error handling, but being related to resource management. i usually use defer to release acquired resources. but there are probably other uses for it!

in languages with try finally constructs, i think of defer as being somewhat equivalent to this:

```
# not go code.

r = acquire_resource() # will raise if it fails to acquire so we wont even enter try block
try {
   do_something_with_resource(r)
   do_something_else_with_resource(r)
} finally {
   release_resource(r)
}
```

similarly to above where try {} block beings immediately after successfully acquiring resource, i argue you want to defer releasing a resource immediately after acquiring it (unless you've got some reason not to , i guess!)

but i guess now i'm talking about error handling with try / finally, as i want to release the resource in all cases, not just in sunny case or error case.


+   it seems like there are a few possible "gotchas" in Go in how go processes exist, and if deferred functions are called or not. e.g. https://golang.org/pkg/os/#Exit